### ISSUE TITLE
How to set optimized options?

### ISSUE THREAD

**Issue by kanryu** [ebiggers/libdeflate#37]
I am currently testing libdeflate to see if png decoding speeds up.

Although libdeflate_deflate_decompress_ex() on libdeflatestatic.lib is executed built using gcc-8 of msys2 in the Windows OS, the execution time is the same as compared with the inflate() of conventional zlib.

Is there an option to make it faster?

I examined the lib file, but there are symbols such as deflate_decompress_bmi.

**Comment by ebiggers** [comment]
It's optimized by default.  Please give details.  What CPU?  How did you measure the speed, exactly?  What input sizes did you test?  Is there any way that I could reproduce your results?  Are you sure it was a controlled test?

**Comment by kanryu** [comment]
I first built it with msvc, but it seems that the optimized function is not included in lib and it is not optimized by this build method. On the other hand, when building with gcc, I confirmed that it contains optimized functions without any special handling.

As for the measurement method of processing, I use lodepng which is one of the OSS implementation of the PNG library, and confirmed it by replacing the expansion part of deflate with libdeflate. Lodepng The development is completed in half the time as it is, but it is almost the same as libpng.

**Comment by ebiggers** [comment]
Yes, due to limited resources to develop and test with MSVC, libdeflate is only properly optimized when built with gcc or clang.

Anyway, if I understand you correctly, your results were:

- lobepng with its own DEFLATE implementation is slow.
- lobepng with libdeflate is fast.
- libpng with zlib is fast.

I don't see where you actually compared libdeflate to zlib directly.  It could be that libpng is faster for other reasons.  Can you please clarify whether you actually did a _controlled_ test that compared libdeflate to zlib?

**Comment by jibsen** [comment]
If you are compiling with GCC for Windows x86 (and not x64), you might also want to add `-msse2`, unless you specifically need to support old machines. Compiling for x86 with SSE2 enabled has been the default in MSVC since 2012.

It is unlikely to affect the decompression speed, since that is handled by inspecting CPU features, but it does appear to speed up compression (as a quick test, the included benchmark tool reports 63 MB/s with the default x86 options, and 76 MB/s with `-msse2` for compression level 1 on silesia.tar).

**Comment by ebiggers** [comment]
Correct, I haven't made the matchfinder optimizations detect CPU features at runtime yet.  So adding
 `-msse2` (for x86) or `-mavx2` (for x86 and x64) will help a bit, if you know the code will only be run on a CPU with those features.  **However**, that only affects compression, whereas the original question here was about decompression.

**Comment by kanryu** [comment]
Hi,
As a result of profiling lodepng from last time, I noticed that adler32 and unfilter processing is late by lodepng 's default (filter has PNG image). Replacing the zlib module with libdeflate instead of inflate and replacing it with unfilter in libpng confirmed that it can be expanded at 1.35 times faster than normal libpng.

It seems necessary to modify the processing of lodepng itself in order to further increase the speed (some unnecessary processing is found).

I added -O3, -Ofast, -mtune etc when adding libdeflate with gcc of msys2 but there was no difference in processing performance.
